### ISSUE TITLE
bad defaulting to match pair?

### ISSUE THREAD

**Issue by macartan** [DeclareDesign/estimatr#260]
I have data like this:

```{r}
> table(dat$TA1, dat$block)
   
     1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17
  0 31  6 31  8  1 17 92 32 73 42 11 15 27 14 50 22  7
  1 11  3 12  3  1  8 24 10 32 10  2  4 23  3 18  7  3
```
And output like this. Not appeared to reduce to matched pair, because of block 5 I suppose, and only 16 degrees of freedom, which seems to suggest just 2 units per block being used...

```
difference_in_means(ncb_wh ~ TA1, blocks = block, data=dat)
 Design:  Matched-pair 
        Estimate Std. Error    t value  Pr(>|t|)   CI Lower CI Upper DF
 TA1 -0.01343043 0.05930286 -0.2264719 0.8237023 -0.1391469 0.112286 16
```

The issue I think is that when it see a block with a matched pair it is concluding that this is a matched pair design. We have discussed this before I know, but surely better behavior is to either drop such small blocks, or combine them with other small blocks.

**Comment by acoppock** [comment]
I think we should probably error in the above case because we don't know how to calculate the standard error.  Agree the output is for sure wrong.

**Comment by macartan** [comment]
Maybe consult with Peter or Cyrus

There are options, eg pool up and include warning. This would be like a
local homoskedacticity assumption I suppose but could imagine being
preferable to tanking. Maybe a force option to allow this.

Could easily imagine this happens by chance with some regularity in some
schemes and choking seems undesirable

**Comment by graemeblair** [comment]
I guess to me choking seems highly informative -- you may learn at the design stage you unintentionally have small blocks and can modify your blocking scheme. At the analysis stage, you may realize you want to change the blocking scheme and our collapsing wouldn't be exactly what you would choose. (In order to select for them, there are many options and I imagine it will be hard in general to select the optimal pooling up algorithm.

If we do allow this, this seems like a surprising behavior -- changing the blocks variable within the estimator! -- so think we should have a more verbose option like force_collapse_blocks to be very clear we are doing this.

**Comment by macartan** [comment]
I see that
I'd be happy with a tank default and force collapse option

**Comment by lukesonnet** [comment]
Okay, implementing fix for the error now and hope to implement the force_collapse_blocks option soon.

**Comment by macartan** [comment]
Not wt machine now but if I remember I got the warning alright but the
behavior was bad. I had all big blocks and only one block with a pair. It
used matched pair which i think meant only two units per blog (which ones?)
and other data seemed to be nit used

**Comment by macartan** [comment]
Apologies if we have gone over before but in mixed cases  is it not
possible to use matched pair approach for all matched pairs and block
approach for larger blocks and then combine variances, given independence
between blocks

**Comment by acoppock** [comment]
My intuition is that that should be correct. Do we need to wait for a proof?

**Comment by lukesonnet** [comment]
Sorry for the delay on this, and thanks for the paper @acoppock.

Macartan, I only mentioned that we had discussed this because I remember coming up with this solution, which I'm sure we agreed at the time was unsatisfactory. I also wanted to ensure you were getting the warning. However, your fear that it is only using two units per block is unfounded. It will apply the classic matched-pairs estimator, which is simply the sample variance of the block-level estimates 
`(1 / (J * (J-1)))   \sum^J_j  (\tauhat_j - \tauhat)^2`

Unfortunately, this is not weighted by the size of your groups, and throws away the precision that could be gained by looking within those larger blocks.

As for your suggestion going forward, have no sense as to whether combining the MP estimator in each paired block and the classical estimator within other blocks would be appropriate. I've never seen that proposed anywhere, and the linked paper doesn't address it directly. That being said, my uncertainty comes from ignorance, and uncertainty about why you can combine two very different within block estimators.

To be clear, what you are proposing is the following, correct?

(1 / (J  * (J - 1)))  \sum^J_j (J * N_j / N)^2 \varhat(\tauhat_j)

where J is the number of blocks, N_j is the number of units in block j, N is the total number of units, and \varhat(\tauhat_j) is the estimated variance within the block which for MP is (\tauhat_j - \tauhat)^2 and for larger blocks is the classic variance estimator using the sample variance for treated and control units.

Is that right? I presume we use that correction (1 / (J * (J-1))) because otherwise there is no correction for using the estimate of tau directly in its variance estimation. That will be "overcorrecting" the variance from the larger blocks but I see no other solution.

As to the paper provided by Alex, this seems like a great general approach, although it seems to be largely speaking finely stratified experiments, where _every_ block has one treatment condition with at most one unit, rather than the case where _some_ blocks have treatment conditions with only one unit. Furthermore, because most of the gains from that estimator are from covariate adjustment, I'm not sure it applies in our case where we don't do covariate adjustment. However, the estimator S^2(Q_1) proposed for the case with groups of varying sizes has several desirable properties and could be tested against:
1) Our current implementation, which is to just get the sample variance of the block estimates (the unweighted matched-pairs estimator)
2) a slightly smarter estimator, which uses the sample variance of the block estimates but re-weights by the size of the blocks
3) the (as far as I can tell) ad-hoc estimator suggested in this thread that I tried to clarify above.

**Comment by macartan** [comment]
Hi Luke -- sorry I haven't gotten to respond to this yet. My thought was to
imagine this as two experiments. One among all those groups with >1 per
stratum (group A) and the other with only 1 per stratum (group B): estimate
\varhat(\tauhat_A) and \varhat(\tauhat_B) using the usual corrections for A
and B and then combine ; now:

 \tauhat_AB = w_A \tauhat_A + w_B \tauhat_B

so

var(\tauhat_AB) = w_A^2 var(\tauhat_A) + w_B^2 var(\tauhat_B)

I believe then that this last step is right; as this is the same as what we
do in the usual blocked approach where we combine the varhats from
different blocks:

varhat(\tauhat_AB) = w_A^2 varhat(\tauhat_A) + w_B^2 varhat(\tauhat_B)

**Comment by lukesonnet** [comment]
No worries Macartan. I just want to clarify:

Group B could have strata with 2 units or even 3+ units, as long as within each strata there is one unit that is alone it its treatment condition. Correct?

And then w_A is defined in the usual way n_A / N, where n_A is the total number of units in all group A strata and N is the total number of units in the sample. Correct?

**Comment by macartan** [comment]
Yes on both (by stratum I meant as defined by block and treatment status).

To be clear though I am just proposing this for discussion. It seems
natural but I have not worked through it

**Comment by acoppock** [comment]
estimatr 2.0.0 no longer falls back to the matched-pair estimator for a design like this one. With exactly one block that has a singleton arm (block 5 here), `difference_in_means()` errors and names the block, because the variance of such blocks is estimated from variation across them. With two or more, larger blocks keep their own within-block variance and the singleton-arm blocks are pooled, following Pashley and Miratrix (2021, https://arxiv.org/abs/1710.10342). The reprex rebuilds the block and arm counts from the table above, with simulated outcomes.

``` r
packageVersion("estimatr")
#> [1] '2.0.0'
library(estimatr)

# Block sizes from the table in the issue: block 5 has one treated and one control unit
n_control <- c(31, 6, 31, 8, 1, 17, 92, 32, 73, 42, 11, 15, 27, 14, 50, 22, 7)
n_treated <- c(11, 3, 12, 3, 1, 8, 24, 10, 32, 10, 2, 4, 23, 3, 18, 7, 3)

set.seed(343)
dat <- data.frame(
  block = rep(rep(1:17, 2), c(n_control, n_treated)),
  TA1 = rep(c(0, 1), c(sum(n_control), sum(n_treated)))
)
dat$Y <- rnorm(nrow(dat)) + 0.1 * dat$TA1
table(dat$TA1, dat$block)
#>    
#>      1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17
#>   0 31  6 31  8  1 17 92 32 73 42 11 15 27 14 50 22  7
#>   1 11  3 12  3  1  8 24 10 32 10  2  4 23  3 18  7  3

# Exactly one block with a singleton arm: an error naming it, not a matched-pair fallback
difference_in_means(Y ~ TA1, blocks = block, data = dat)
#> Error in `blocked_variance_pm()`:
#> ! Only one block has a single treated or control unit (block 5).
#> The variance contributed by such blocks is estimated from the variation across them, so at least two are needed.
#> Merge that block with another, drop it, or use `lm_robust()` with block fixed effects.

# Merging block 5 into block 4 gives the ordinary blocked estimator on all units
dat$block_merged <- ifelse(dat$block == 5, 4, dat$block)
difference_in_means(Y ~ TA1, blocks = block_merged, data = dat)
#> Design:  Blocked 
#>       Estimate Std. Error   t value  Pr(>|t|)   CI Lower  CI Upper  DF
#> TA1 0.07786397 0.09460106 0.8230771 0.4107801 -0.1079128 0.2636407 621

# With a second singleton-arm block, the small blocks are pooled (Pashley and Miratrix 2021)
extra <- data.frame(block = 18, TA1 = c(0, 1), Y = rnorm(2))
dat2 <- rbind(dat[, c("block", "TA1", "Y")], extra)
difference_in_means(Y ~ TA1, blocks = block, data = dat2)
#> Design:  Hybrid blocked 
#>       Estimate Std. Error   t value  Pr(>|t|)  CI Lower  CI Upper       DF
#> TA1 0.08218729   0.094325 0.8713204 0.3839166 -0.103048 0.2674226 619.8255
```

The fix is in estimatr 2.0.0, which is on `main` and has been submitted to CRAN; until CRAN accepts it, `remotes::install_github("DeclareDesign/estimatr")` installs it.
